temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+                    # _run_activity when the activity function is obtained
+                    activity = _RunningActivity(
+                        task=None,
+                        cancelled_event=temporalio.activity._CompositeEvent(
+                            thread_event=threading.Event(), async_event=None
+                        ),
+                        worker_shutdown_event=temporalio.activity._CompositeEvent(
+                            thread_event=threading.Event(), async_event=None
+                        ),
+                        sync=False,
+                        last_heartbeat_task=None,
+                    )
+                    activity.task = asyncio.create_task(
+                        self._run_activity(task.task_token, task.start, activity)
+                    )
+                    self._running_activities[task.task_token] = activity
+                elif task.HasField("cancel"):
+                    self._cancel_activity(task.task_token, task.cancel)
+                else:
+                    logger.warning("unrecognized activity task: %s", task)
+            except temporalio.bridge.worker.PollShutdownError:
+                return
+            except Exception as err:
+                # Should never happen
+                logger.exception(f"Activity runner failed: {err}", extra={"error": err})
+
+    async def _run_workflows(self) -> None:
+        raise NotImplementedError
+
+    def _cancel_activity(
+        self, task_token: bytes, cancel: temporalio.bridge.proto.activity_task.Cancel
+    ) -> None:
+        activity = self._running_activities.get(task_token)
+        if not activity:
+            logger.warning("Cannot find activity to cancel for token %s", task_token)
+            return
+        logger.debug("Cancelling activity %s, reason: %s", task_token, cancel.reason)
+        activity.cancel(cancelled_by_request=True)
+
+    def _heartbeat_activity(self, task_token: bytes, *details: Any) -> None:
+        # We intentionally make heartbeating non-async, but since the data
+        # converter is async, we have to schedule it
+        logger = temporalio.activity.logger
+        activity = self._running_activities.get(task_token)
+        if activity:
+            activity.current_heartbeat_seq += 1
+            heartbeat_seq = activity.current_heartbeat_seq
+            activity.last_heartbeat_task = asyncio.create_task(
+                self._heartbeat_activity_async(
+                    logger, task_token, heartbeat_seq, *details
+                )
+            )
+
+    async def _heartbeat_activity_async(
+        self,
+        logger: logging.Logger,
+        task_token: bytes,
+        heartbeat_seq: int,
+        *details: Any,
+    ) -> None:
+        activity = self._running_activities.get(task_token)
+        # Bail if not the latest
+        if not activity or activity.current_heartbeat_seq != heartbeat_seq:
+            return
```

Comment:
How would that happen?

Reply:
> If we bail here we might never be able to emit any heartbeats if activity heartbeats too fast.

Not necessarily here, but yes if we continue to bail after heartbeat. Simple buffering is rough on memory, but I can do something similar. The problem is I have to implement advanced ordering in case the data conversion for the most recent heartbeat is slower than the data conversion for an earlier one.

What happens in TS if my activity does:

```typescript
for {
  Context.current().heartbeat()
}
```

Also, what happens in TypeScript if a later heartbeat's data conversion finishes after an earlier heartbeat's?

Reply:
We need to improve TS heartbeats regardless of what we do here.
Instead of a list you can just update the last value somewhere.

Reply:
:+1: I will change from seq to maintaining a current and next-pending heartbeat so that any heartbeat that "starts" will properly complete.

Reply:
Changed up the heartbeating to maintain a current and pending heartbeat (w/ pending repeatedly replaced while one is running).